### DISCUSSION TITLE
svgAsPngFile and unsafePerformIO

### DISCUSSION THREAD

**Discussion by nbrader** [reanimate/reanimate#210] [Ideas]
A nice looking package but I was wondering why you have

`svgAsPngFile :: SVG -> FilePath`

instead of

`svgAsPngFile :: SVG -> IO FilePath`

**Comment by lemmih** [comment]
From the point of view of your animation, the `FilePath` is purely determined by the SVG. The same input will always give the same output. There's no need for it to be IO since it doesn't matter when it is called or how many times it is executed. It'll give the same answer no matter what.

**Reply by hasufell** [reply]
That makes no sense. You're not just returning a filepath, you're

1. writing stuff to disk
2. doing stuff with temporary files, rename, copy, ...
3. non-atomically check for file existence
4. invoke inkscape/rsvg-convert etc

All of these can fail in a bazillion ways and depend on filesystem/system state.

Now your users can't catch any of those errors reasonably, because of lazy IO.

----

> It'll give the same answer no matter what.

It definitely won't. A few simple reasons why:

1. permission skews on temp directories -> failure
2. filesystem full -> failure
3. inkscape broken/non-existent or simply invoked with incorrect arguments, because of diverging versions -> failure

But those are only a few I could come up with off the top of my head. There'll be many more.

`unsafePerformIO` is **not** a design pattern. It's used for pure FFI calls and very nasty performance optimizations, not to signal "the output should be the same".

**Reply by lemmih** [reply]
> Now your users can't catch any of those errors reasonably, because of lazy IO.

There _is_ no reasonable way to handle those errors without aborting the animation. You're free to handle errors at the top-level but you'll have to restart the animation process.

> > It'll give the same answer no matter what.
> 
> It definitely won't. A few simple reasons why:

It'll either give the same answer or it'll not return.

All Haskell functions might fail. Imagine you've run out of heap and then try to add two integers. Addition is completely pure but it can still fail if heap allocation fails. No need for it to be in IO, though: Either the addition will succeed and give you the right result or it will fail and give no result at all.

I'd love to hear a reason for why 'unsafePerformIO' shouldn't be used other than "it might fail."

**Reply by hasufell** [reply]
> All Haskell functions might fail

I think you're confusing evaluation and execution and what **purity** really means: weak equivalence of call-by-need, call-by-value and call-by-name (see http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.27.7800&rep=rep1&type=pdf). Purity isn't even defined for execution, only for evaluation. All IO actions are pure, because they do nothing when evaluated.

What `unsafePerformIO` does is break this language guarantee at the discretion of the programmer: now evaluation of a function may yield different results depending on the evaluation strategy.

> I'd love to hear a reason for why 'unsafePerformIO' shouldn't be used other than "it might fail."

1. it breaks purity
2. the types are lying now and convey misinformation to the user
3. it makes it incredibly hard for the user to catch underlying exceptions, now they have to utilize things like `evaluate` and `force` etc otherwise exceptions may randomly leak through exception handlers because of lazy IO
4. depending on filesystem operations and filesystem state is **always**  IO and can **always** fail. The filesystem is your **input**.

> Imagine you've run out of heap

That's a pretty bad argument. Why not use unsafePerformIO everywhere then? Because you'll not be able to reason about your program anymore.

**Reply by lemmih** [reply]
> > All Haskell functions might fail
> 
> I think you're confusing evaluation and execution and what **purity** really means: weak equivalence of call-by-need, call-by-value and call-by-name (see http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.27.7800&rep=rep1&type=pdf). Purity isn't even defined for execution, only for evaluation. All IO actions are pure, because they do nothing when evaluated.
> 
> What `unsafePerformIO` does is break this language guarantee at the discretion of the programmer: now evaluation of a function may yield different results depending on the evaluation strategy.
> 
> > I'd love to hear a reason for why 'unsafePerformIO' shouldn't be used other than "it might fail."
> 
> 1. it breaks purity

It doesn't always. The `svgAsPngFile` function returns values that are identical for identical arguments, and it has no observable side-effects. It is a pure function.

> 2. the types are lying now and convey misinformation to the user

This is subjective. I think you're wrong but you're entitled to this belief.

> 3. it makes it incredibly hard for the user to catch underlying exceptions, now they have to utilize things like `evaluate` and `force` etc otherwise exceptions may randomly leak through exception handlers because of lazy IO

This is incorrect. The `reanimate :: Animation -> IO ()` function will either succeed or raise an exception. You're free to handle them at this point.

You're right that catching exception in pure code is incredibly hard. This is because you absolutely should not do it.

> 4. depending on filesystem operations and filesystem state is **always**  IO and can **always** fail
> 
> > Imagine you've run out of heap
> 
> That's a pretty bad argument. Why not use unsafePerformIO everywhere then? Because you'll not be able to reason about your program anymore.

Some operations _do_ need to be sequenced. Changing the order of `readMVar` and `modifyMVar` will change observable behaviour and you therefore shouldn't use `unsafePerformIO`. But `svgAsPngFile` does not care when it is called or how many times it is called. It will always give you the same result (or fail).

I believe the crux of our disagreement is that you think `svgAsPngFile` is not a pure function while I think it is.

**Reply by chowells79** [reply]
> The same input will always give the same output.

This is actually false. I note that the code depends on `getXdgDirectory`.

```
Prelude System.Directory System.Environment> getXdgDirectory XdgCache "foo"
"/home/carl/.cache/foo"
Prelude System.Directory System.Environment> setEnv "XDG_CACHE_HOME" "/mars"
Prelude System.Directory System.Environment> getXdgDirectory XdgCache "foo"
"/mars/foo"
```

Why might someone do that?  Who knows.  It's their business.  But it means that the type, like the cake, is a lie.

**Reply by lemmih** [reply]
That is absolutely true. Nice catch. The cache directory should be cached at the start of a render such that `svgAsPngFile` is truly a pure function. Fixing this is not high on my TODO list, though.

**Reply by chowells79** [reply]
Putting IO in the type would be a much easier fix, and also prevent other shenanigans of that sort.

**Comment by nbrader** [comment]
Hi David. I'm the original poster and I wanted to share my perspective.

I heard about your project when it was mentioned in a recent video by the YouTube channel "3Blue1Brown" (link to video: https://youtu.be/ojjzXyQCzso?t=1157). I happen to have a script written in Haskell which could use some graphics so this piqued my interest. For me, graphical output has always made programming that much more rewarding so I'm glad people like you help broaden the tool-set in Haskell.

I won't pretend to be anything like an expert at writing Haskell code, despite dabbling in it for a number of years now. You've probably already achieved far more in this one package than I've achieved in all the numerous scripts I've written.

I posted the question because the lack of the IO monad in a clearly effectful function surprised me. When you responded I thought it was worth asking for second opinion on the #haskell irc channel on https://libera.chat/. What followed in this thread was the result of that.

So for what it's worth, it seems to me you're using a nonstandard definition of "purity", at least in contrast with it's use in Prelude. To borrow an example given in that IRC chat, putStrLn :: String -> IO () but by your reasoning wouldn't this be putStrLn :: String -> (). svgAsPngFile is but one of a number of your functions that you had to define using performUnsafeIO in order to do the very basic IO operation of writing a file. I would have taken this as a hint that I was working against the intended use of the language.

I can only imagine more could be said by people who understand better than me what optimizations the compiler does with non-IO values and how this use of performUnsafeIO does or does not break it's assumptions. The most I can do is point to https://hackage.haskell.org/package/base-4.12.0.0/docs/System-IO-Unsafe.html#v:unsafePerformIO which seems to list a few ways of ensuring it's use is safe.

As for the graphics I'm hoping to produce, I'm hoping this package or possibly the diagrams package (https://hackage.haskell.org/package/diagrams) can help me.

I hope you can see where I'm coming from and correct me if I've got the wrong idea about any of that.

**Reply by lemmih** [reply]
> Hi David. I'm the original poster and I wanted to share my perspective.

Hi nbrader. I'm glad you raised the original question. It's quite an interesting topic.

> I heard about your project when it was mentioned in a recent video by the YouTube channel "3Blue1Brown" (link to video: https://youtu.be/ojjzXyQCzso?t=1157). I happen to have a script written in Haskell which could use some graphics so this piqued my interest. For me, graphical output has always made programming that much more rewarding so I'm glad people like you help broaden the tool-set in Haskell.
> 
> I won't pretend to be anything like an expert at writing Haskell code, despite dabbling in it for a number of years now. You've probably already achieved far more in this one package than I've achieved in all the numerous scripts I've written.
> 
> I posted the question because the lack of the IO monad in a clearly effectful function surprised me.

Well, the crux of the matter is what should be considered an observable side-effect. From one point of view, writing a PNG image to a file is clearly a side-effect. But it's all a matter of perspective: Imagine we had a magic hard-drive containing all possible PNG images. Now the `svgAsPngFile` doesn't have any side-effects and it merely returns the path to a file that was already there.

So, `svgAsPngFile` is referentially transparent and doesn't change anything that is observable from your animation. As such, it has no business in the IO monad.

> When you responded I thought it was worth asking for second opinion on the #haskell irc channel on https://libera.chat/. What followed in this thread was the result of that.
> 
> So for what it's worth, it seems to me you're using a nonstandard definition of "purity", at least in contrast with it's use in Prelude. To borrow an example given in that IRC chat, putStrLn :: String -> IO () but by your reasoning wouldn't this be putStrLn :: String -> (). svgAsPngFile is but one of a number of your functions that you had to define using performUnsafeIO in order to do the very basic IO operation of writing a file. I would have taken this as a hint that I was working against the intended use of the language.

I think we're talking past each other. If `putStrLn` had the type `String -> ()` then that clearly would cause problems. But I don't see how `svgAsPngFile` can ever cause problems due to not being in the IO monad. If you can come up with an example, I'm all ears.

> I can only imagine more could be said by people who understand better than me what optimizations the compiler does with non-IO values and how this use of performUnsafeIO does or does not break it's assumptions. The most I can do is point to https://hackage.haskell.org/package/base-4.12.0.0/docs/System-IO-Unsafe.html#v:unsafePerformIO which seems to list a few ways of ensuring it's use is safe.

Reanimate's use of `unsafePerformIO` is safe. As its name suggest, though, it can absolutely be used unsafely so I understand your mistrust. I started working full-time as a Haskell developer 14 years ago, I've worked on both GHC and the LHC Haskell compilers, and I promise you that I'm using `unsafePerformIO` safely.

> As for the graphics I'm hoping to produce, I'm hoping this package or possibly the diagrams package (https://hackage.haskell.org/package/diagrams) can help me.

Great. The diagrams library is a jewel. I had good integration with it in the past but the code was too difficult to maintain, unfortunately.

**Comment by madnight** [comment]
@lemmih 

I've talked with many Haskellers over the years and your standpoint is really different to what I'm used to hear on IRC or reddit.

```
-- | Convert an SVG object to a pixel-based image and save it to disk, returning
--   the filepath. The default resolution is 2560x1440. See also 'svgAsPngFile''.
--   Multiple raster engines are supported and are selected using the '--raster'
--   flag in the driver.
svgAsPngFile :: SVG -> FilePath
```

From your comment alone I would immediately expect the function return type to be of type IO. And feel very confused about your promise that this is actually is referentially transparent, especially when @chowells79 already found a flaw in that argument and points out that `getXdgDirectory` in combination with unsafePerformIO already breaks referentially transparency, which is no surprise to me.

I would suggest to you to talk to some other Haskellers you might know and get their opinion about the matter. To me it looks like a really strange decision and I don't buy that it doesn't break referentially transparency. Anyway I really like your project, so thanks for that one.

**Reply by lemmih** [reply]
I find your comment to be non-constructive. As someone who has been around for a while (I'm the original author of cabal-install so I've been around long enough that you've probably used my software in one way or another), please accept that am well aware not everyone agrees with my choices.

If you have a bug report, a feature request, a PR, or even if you want to be a co-maintainer, I'm all ears.
